File: clases/clase1/tarea/pacmankata/PacManKata/Maze.cs
using System;

namespace PacManKata
{
    public class Maze
    {
        private class PacManPosition
        {
            public int Row { get; set; }
            public int Column { get; set; }

            public PacManPosition(int row, int column)
            {
                this.Row = row;
                this.Column = column;
            }
        }

        private enum PacManDirections
        {
            Up = 1,
            Right,
            Down,
            Left
        }

        private int rows;
        private int columns;

        private PacManPosition pacManPosition;
        private PacManDirections pacManDirection;        

        public Maze(int rows, int columns)
        {
            this.rows = rows;
            this.columns = columns;

            int middleRow = this.rows % 2 == 0? this.rows / 2 : (this.rows + 1) / 2;
            int middleColumn = this.columns % 2 == 0? this.columns / 2 : (this.columns+ 1) / 2;
            this.pacManPosition = new PacManPosition(middleRow, middleColumn);
            this.pacManDirection = PacManDirections.Up;
        }

        public int Rows { get { return rows; } }
        public int Columns { get { return columns; } }

        public bool IsPacManAt(int row, int column)
        {
            return pacManPosition.Row == row && pacManPosition.Column == column;
        }

        public bool IsPacManLookingDown()
        {
            return pacManDirection == PacManDirections.Down;
        }

        public bool IsPacManLookingUp()
        {
            return pacManDirection == PacManDirections.Up;
        }

        public bool IsPacManLookingLeft()
        {
            return pacManDirection == PacManDirections.Left;
        }

        public bool IsPacManLookingRight()
        {
            return pacManDirection == PacManDirections.Right;
        }

        public bool IsEmptyAt(int row, int column)
        {
            return !IsPacManAt(row, column);
        }

        public void PacManDown()
        {
            pacManDirection = PacManDirections.Down;
        }

        public void PacManUp()
        {
            pacManDirection = PacManDirections.Up;
        }

        public void PacManLeft()
        {
            pacManDirection = PacManDirections.Left;
        }

        public void PacManRight()
        {
            pacManDirection = PacManDirections.Right;
        }

        public void Tick()
        {
            switch (pacManDirection)
            {
                case PacManDirections.Up:
                    pacManPosition.Row--;
                    ValidateRowPosition();
                    break;
                case PacManDirections.Down:
                    pacManPosition.Row++;
                    ValidateRowPosition();
                    break;
                case PacManDirections.Right:
                    pacManPosition.Column++;
                    ValidateColumnPosition();
                    break;                
                case PacManDirections.Left:
                    pacManPosition.Column--;
                    ValidateColumnPosition();
                    break;
            }
        }

        private void ValidateRowPosition()
        {
            if (pacManPosition.Row > this.rows)
            {
                pacManPosition.Row = 1;
            }
            else if (pacManPosition.Row < 1)
            {
                pacManPosition.Row = this.rows;
            }
        }

        private void ValidateColumnPosition()
        {
            if (pacManPosition.Column > this.columns)
            {
                pacManPosition.Column = 1;
            }
            else if (pacManPosition.Column < 1)
            {
                pacManPosition.Column = this.columns;
            }
        }

        public void Print()
        {
        }
    }    
}

File: javascript/PacMan/Maze.js
function Maze(rows, columns){
	let PacManDirections = { up: 1, right: 2, down: 3, left: 4 };
	let middleRow = rows % 2 === 0? rows / 2 : (rows + 1) / 2;
    let middleColumn = columns % 2 === 0? columns / 2 : (columns+ 1) / 2;

	this.rows = rows;
	this.columns = columns;
	this.pacManPosition = { row: middleRow, column: middleColumn };	
    this.pacManDirection = PacManDirections.up;

	this.isPacManAt = (row, column) => this.pacManPosition.row == row && this.pacManPosition.column == column;
	this.isPacManLookingDown = () => this.pacManDirection == PacManDirections.down;
	this.isPacManLookingUp = () => this.pacManDirection == PacManDirections.up;
	this.isPacManLookingLeft = () => this.pacManDirection == PacManDirections.left;
	this.isPacManLookingRight = () => this.pacManDirection == PacManDirections.right;
	this.isEmptyAt = (row, column) => !this.isPacManAt(row, column);
	this.pacManDown = () => this.pacManDirection = PacManDirections.down;
	this.pacManUp = () => this.pacManDirection = PacManDirections.up;
	this.pacManLeft = () => this.pacManDirection = PacManDirections.left;
	this.pacManRight = () => this.pacManDirection = PacManDirections.right;
	this.tick = doTick;
	this.print = () => {};

	function doTick(){
		switch(this.pacManDirection){
			case PacManDirections.up:
                this.pacManPosition.row--;
                validatePosition('row','rows');
                break;
            case PacManDirections.down:
                this.pacManPosition.row++;
                validatePosition('row','rows');
                break;
            case PacManDirections.right:
                this.pacManPosition.column++;
                validatePosition('column', 'columns');
                break;                
            case PacManDirections.left:
                this.pacManPosition.column--;
                validatePosition('column', 'columns');
                break;
		}
	}

	let validatePosition = (pacManProp, mazeProp) => {
        if (this.pacManPosition[pacManProp] > this[mazeProp])
        {
            this.pacManPosition[pacManProp] = 1;
        }
        else if (this.pacManPosition[pacManProp] < 1)
        {
            this.pacManPosition[pacManProp] = this[mazeProp];
        }
    }	
}